File: addons/guide-cpp/src/modifiers/guide_modifier_curve.cpp
#include "guide_modifier_curve.h"
#include <godot_cpp/variant/utility_functions.hpp>

using namespace godot;

void GUIDEModifierCurve::_bind_methods() {
    ClassDB::bind_method(D_METHOD("get_curve"), &GUIDEModifierCurve::get_curve);
    ClassDB::bind_method(D_METHOD("set_curve", "curve"), &GUIDEModifierCurve::set_curve);
    ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "curve", PROPERTY_HINT_RESOURCE_TYPE, "Curve"), "set_curve", "get_curve");

    ClassDB::bind_method(D_METHOD("get_x"), &GUIDEModifierCurve::get_x);
    ClassDB::bind_method(D_METHOD("set_x", "val"), &GUIDEModifierCurve::set_x);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "x"), "set_x", "get_x");

    ClassDB::bind_method(D_METHOD("get_y"), &GUIDEModifierCurve::get_y);
    ClassDB::bind_method(D_METHOD("set_y", "val"), &GUIDEModifierCurve::set_y);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "y"), "set_y", "get_y");

    ClassDB::bind_method(D_METHOD("get_z"), &GUIDEModifierCurve::get_z);
    ClassDB::bind_method(D_METHOD("set_z", "val"), &GUIDEModifierCurve::set_z);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "z"), "set_z", "get_z");
}

GUIDEModifierCurve::GUIDEModifierCurve() {
    curve.instantiate();
    curve->add_point(Vector2(0, 0));
    curve->add_point(Vector2(1, 1));
}

Vector3 GUIDEModifierCurve::_modify_input(Vector3 input, double delta, int value_type) const {
    if (curve.is_null()) {
        UtilityFunctions::push_error("No curve added to Curve modifier.");
        return input;
    }
    if (!input.is_finite()) return Vector3(Math_INF, Math_INF, Math_INF);
    return Vector3(
        x ? curve->sample(input.x) : input.x,
        y ? curve->sample(input.y) : input.y,
        z ? curve->sample(input.z) : input.z
    );
}

bool GUIDEModifierCurve::is_same_as(const Ref<GUIDEModifier> &other) const {
    Ref<GUIDEModifierCurve> o = other;
    if (o.is_null()) return false;
    return o->get_curve() == curve && 
           o->get_x() == x && 
           o->get_y() == y && 
           o->get_z() == z;
}

File: addons/GuideCS/Modifiers/GUIDEModifierCurve.cs
using Godot;

namespace Guide;

[Tool]
[GlobalClass]
public partial class GUIDEModifierCurve : GUIDEModifier {
	[Export] public Curve Curve { get; set; } = DefaultCurve();
	[Export] public bool X { get; set; } = true;
	[Export] public bool Y { get; set; } = true;
	[Export] public bool Z { get; set; } = true;

	public static Curve DefaultCurve() {
		var curve = new Curve();
		curve.AddPoint(new Vector2(0.0f, 0.0f));
		curve.AddPoint(new Vector2(1.0f, 1.0f));
		return curve;
	}

	public override bool IsSameAs(GUIDEModifier other) {
		return other is GUIDEModifierCurve otherCurve &&
			   Curve == otherCurve.Curve &&
			   X == otherCurve.X &&
			   Y == otherCurve.Y &&
			   Z == otherCurve.Z;
	}

	public override Vector3 _ModifyInput(Vector3 input, float delta, GUIDEAction.GUIDEActionValueType valueType) {
		if (Curve == null) {
			GD.PushError("No curve added to Curve modifier.");
			return input;
		}

		if (!input.IsFinite()) return Vector3.Inf;

		return new Vector3(
			X ? Curve.Sample(input.X) : input.X,
			Y ? Curve.Sample(input.Y) : input.Y,
			Z ? Curve.Sample(input.Z) : input.Z
		);
	}

	public override string _EditorName() => "Curve";

	public override string _EditorDescription() => "Applies a curve to each input axis.";
}
